cw_decred/lib/wallet.dart

```diff
@@ -0,0 +1,604 @@
+    // DiscoveringAddrs = 3
+    // Rescanning = 4
+    // Complete = 5
+
+    if (syncStatusCode > 4) {
+      syncStatus = SyncedSyncStatus();
+      // Initiate a receive address in case we lose peers later.
+      if (walletAddresses.currentAddr == '') {
+        walletAddresses.address;
+      }
+      return true;
+    }
+
+    if (syncStatusCode == 0) {
+      syncStatus = ConnectedSyncStatus();
+      return false;
+    }
+
+    if (syncStatusCode == 1) {
+      syncStatus = SyncingSyncStatus(targetHeight, 0.0);
+      return false;
+    }
+
+    if (syncStatusCode == 2) {
+      final headersProg = headersHeight / targetHeight;
+      // Only allow headers progress to go up half way.
+      syncStatus =
+          SyncingSyncStatus(targetHeight - headersHeight, headersProg / 2);
+      return false;
+    }
+
+    // TODO: This step takes a while so should really get more info to the UI
+    // that we are discovering addresses.
+    if (syncStatusCode == 3) {
+      // Hover at half.
+      syncStatus = SyncingSyncStatus(0, .5);
+      return false;
+    }
+
+    if (syncStatusCode == 4) {
+      // Start at 75%.
+      final rescanProg = rescanHeight / targetHeight / 4;
+      syncStatus =
+          SyncingSyncStatus(targetHeight - rescanHeight, .75 + rescanProg);
+      return false;
+    }
+    return false;
+  }
+
+  @action
+  @override
+  Future<void> connectToNode({required Node node}) async {
+    if (connecting) {
+      throw "decred already connecting";
+    }
+    connecting = true;
+    String addr = "";
+    if (node.uri.host != "") {
+      addr = node.uri.host;
+      if (node.uri.port != "") {
+        addr += ":" + node.uri.port.toString();
+      }
+    }
+    if (addr != persistantPeer) {
+      if (syncTimer != null) {
+        syncTimer!.cancel();
+        syncTimer = null;
+      }
+      persistantPeer = addr;
+      libdcrwallet.closeWallet(walletInfo.name);
+      final network = walletInfo.derivationPath ==
+                  DecredWalletService.seedRestorePathTestnet ||
+              walletInfo.derivationPath ==
+                  DecredWalletService.pubkeyRestorePathTestnet
+          ? "testnet"
+          : "mainnet";
+      libdcrwallet.loadWalletSync({
+        "name": walletInfo.name,
+        "dataDir": walletInfo.dirPath,
+        "network": network,
+      });
+    }
+    await this._startSync();
+    connecting = false;
+  }
+
+  @action
+  @override
+  Future<void> startSync() async {
+    if (connecting) {
+      throw "decred already connecting";
+    }
+    connecting = true;
+    await this._startSync();
+    connecting = false;
+  }
+
+  Future<void> _startSync() async {
+    if (syncTimer != null) {
+      return;
+    }
+    try {
+      syncStatus = ConnectingSyncStatus();
+      libdcrwallet.startSyncAsync(
+        name: walletInfo.name,
+        peers: persistantPeer,
+      );
+      syncTimer = Timer.periodic(
+          Duration(seconds: 5), (Timer t) => performBackgroundTasks());
```

Comment:
It is only fetching until it sees one it recognizes, not all. Batches of five.

Five seconds has been good for testing on simnet, but yeah can be whatever you think is best. Setting to 30 seconds.

Reply:
I get that it's some batches of five, but you still have it in a while true, so it will keep going until it fetches all transactions, which will repeat every 5 seconds, since you set `from` to 0 every time

so why not just save the fetched transactions, and save the last scanned transaction count, and then just try and fetch from that count if you don't find any new transactions it will break out of the loop on the first time

Reply:
I think this is why I originally was saving the block hash so we could skip checking if the block hadn't changed, then I realized that we also want to show pending transactions in the mempool.

Unless I coded something wrong, it should only fetch until it sees a known transaction every time, from the newest, so it would stop at the first batch most of the time. 0 is not a block number if that's what you mean, which is probably where the confusion is from. `this.transactionHistory.update` will return true and the function will return if there is an old tx.

from is the number of transactions to skip, which is not obvious.
```
$ ./dcrctl help listtransactions
listtransactions ("account" count=10 from=0 includewatchonly=false)

Returns a JSON array of objects containing verbose details for wallet transactions.

Arguments:
1. account          (string, optional)                 DEPRECATED -- Unused (must be unset or "*")
2. count            (numeric, optional, default=10)    Maximum number of transactions to create results from
3. from             (numeric, optional, default=0)     Number of transactions to skip before results are created
4. includewatchonly (boolean, optional, default=false) Unused
```

Also this is only a db check, it is not querying other nodes with this call. I do not think checking a number in the db will be faster. Currently this would also mean we have to fetch all transactions every time since I don't think wallet has a method to tell us how many tx there are. I can imagine such a change being met with resistance as well since it doesn't do much.